File: terminal/claude/hooks/auto-improve.py
#!/usr/bin/env python3

import fcntl
import hashlib
import json
import os
from pathlib import Path
import sys
import tempfile


REVIEW = (
    "Load the auto-improve skill for this automatic, one-per-session checkpoint. "
    "Inspect the current session and relevant setup files read-only. Show at most "
    "3 concrete, evidence-backed proposals for skills, hooks, scripts, or rules, "
    "and ask which to apply. Make no setup or code edits, commits, or external "
    "calls to publish. If no proposal is useful, finish quietly. Do not issue "
    "another automatic review in this session."
)


def valid_state(state):
    if not isinstance(state, dict) or set(state) != {
        "current_prompt_id", "completed_prompt_ids", "issued"
    }:
        return False
    completed = state["completed_prompt_ids"]
    current = state["current_prompt_id"]
    if not isinstance(completed, list) or len(completed) > 3:
        return False
    ids = completed + ([] if current is None else [current])
    return (
        all(
            isinstance(value, str)
            and len(value) == 64
            and all(char in "0123456789abcdef" for char in value)
            for value in ids
        )
        and len(set(ids)) == len(ids)
        and type(state["issued"]) is bool
        and state["issued"] == (len(completed) == 3)
        and (not state["issued"] or current is None)
    )


def main():
    phase = "input"
    try:
        event = json.load(sys.stdin)
        if not isinstance(event, dict):
            raise ValueError("invalid event")
        name = event.get("hook_event_name")
        if name not in ("UserPromptSubmit", "Stop") or "agent_id" in event:
            return
        session_id = event.get("session_id")
        prompt_id = event.get("prompt_id")
        if not session_id or not prompt_id:
            return
        if not isinstance(session_id, str) or not isinstance(prompt_id, str):
            raise ValueError("invalid IDs")
        active = event.get("stop_hook_active", False)
        if type(active) is not bool:
            raise ValueError("invalid continuation flag")
        if active:
            return

        phase = "state"
        os.umask(0o077)
        root = Path(os.environ.get("XDG_STATE_HOME") or Path.home() / ".local/state")
        if not root.is_absolute():
            raise ValueError("state root must be absolute")
        directory = root / "auto-improve/claude"
        directory.mkdir(parents=True, exist_ok=True, mode=0o700)
        session_key = hashlib.sha256(session_id.encode()).hexdigest()
        prompt_key = hashlib.sha256(prompt_id.encode()).hexdigest()
        state_path = directory / (session_key + ".json")
        with (directory / (session_key + ".lock")).open("a") as lock:
            fcntl.flock(lock, fcntl.LOCK_EX | fcntl.LOCK_NB)
            if state_path.exists() or state_path.is_symlink():
                with state_path.open() as source:
                    state = json.load(source)
                if not valid_state(state):
                    raise ValueError("invalid state")
            else:
                if name == "Stop":
                    return
                state = {
                    "current_prompt_id": None,
                    "completed_prompt_ids": [],
                    "issued": False,
                }
            if state["issued"] or prompt_key in state["completed_prompt_ids"]:
                return
            if name == "UserPromptSubmit":
                if state["current_prompt_id"] == prompt_key:
                    return
                state["current_prompt_id"] = prompt_key
            else:
                if state["current_prompt_id"] != prompt_key:
                    return
                state["current_prompt_id"] = None
                state["completed_prompt_ids"].append(prompt_key)
                state["issued"] = len(state["completed_prompt_ids"]) == 3

            temporary = None
            try:
                with tempfile.NamedTemporaryFile(
                    mode="w", dir=directory, prefix=session_key + ".",
                    suffix=".tmp", delete=False,
                ) as target:
                    temporary = Path(target.name)
                    json.dump(state, target)
                    target.flush()
                    os.fsync(target.fileno())
                os.replace(temporary, state_path)
                directory_fd = os.open(directory, os.O_RDONLY)
                try:
                    os.fsync(directory_fd)
                finally:
                    os.close(directory_fd)
            finally:
                if temporary is not None:
                    temporary.unlink(missing_ok=True)

            # Persist first: a crash before output may skip a review, never repeat it.
            if state["issued"]:
                phase = "output"
                print(json.dumps({"hookSpecificOutput": {
                    "hookEventName": "Stop", "additionalContext": REVIEW,
                }}), flush=True)
    except (OSError, ValueError, RuntimeError) as error:
        print(
            f"auto-improve: {phase} failed ({type(error).__name__}); no review.",
            file=sys.stderr,
        )


if __name__ == "__main__":
    main()

File: terminal/claude/hooks/auto-improve.ts
#!/usr/bin/env node

import { spawn } from "node:child_process";
import { createHash, randomBytes, randomUUID } from "node:crypto";
import {
  closeSync,
  fsyncSync,
  lstatSync,
  mkdirSync,
  openSync,
  readFileSync,
  renameSync,
  rmSync,
  writeFileSync,
  writeSync,
} from "node:fs";
import { homedir } from "node:os";
import { isAbsolute, join } from "node:path";

// Positional args only ($1..$7, plus the inherited $DOTFILES_PATH env var):
// never interpolated, so a prompt or path can hold anything without
// touching the script text.
const REVIEWER_SCRIPT = `
dir=$1
uuid=$2
session_id=$3
cwd=$4
transcript=$5
date=$6
prompt=$7
tmp="$dir/$uuid.report.md.tmp"
err="$dir/$uuid.err"
{
  echo "# auto-improve report"
  echo "source_session: $session_id"
  echo "cwd: $cwd"
  echo "transcript: $transcript"
  echo "resume: claude --resume $uuid"
  echo "date: $date"
  echo ""
} > "$tmp"
if [ -n "$DOTFILES_PATH" ]; then
  claude -p "$prompt" --session-id "$uuid" --permission-mode plan --add-dir "$DOTFILES_PATH" >> "$tmp" 2> "$err"
else
  claude -p "$prompt" --session-id "$uuid" --permission-mode plan >> "$tmp" 2> "$err"
fi
status=$?
if [ "$status" -eq 0 ]; then
  [ -s "$err" ] || rm -f "$err"
  mv "$tmp" "$dir/$uuid.report.md"
else
  cat "$err" >> "$tmp"
  rm -f "$err"
  mv "$tmp" "$dir/$uuid.failed.md"
fi
`;

function buildPrompt(transcriptPath: string, cwd: string): string {
  return (
    "Automatic, one-per-session auto-improve checkpoint for another Claude " +
    "Code session. Load the auto-improve skill with the Skill tool and " +
    `follow it. Evidence: that session's transcript at ${transcriptPath} ` +
    "(JSONL; read all of it, and only that transcript) with working " +
    `directory ${cwd}. Inspect the relevant skills, hooks, scripts, and ` +
    "rules read-only. Do the work yourself without subagents. Return the " +
    "concrete, evidence-backed proposals the transcript supports, strongest " +
    "first, for skills, hooks, scripts, or rules, each with its evidence, " +
    "target path, change, and verification, or state that none is useful. " +
    "This run is read-only: " +
    "make no edits, commits, or external calls. Plan mode is only the " +
    "read-only guard: do not write a plan file, do not start Explore or Plan " +
    "agents, and do not call ExitPlanMode; answer with the proposals " +
    "directly. The user may resume this session later; apply a proposal " +
    "only when they ask for it then."
  );
}

type State = {
  current_prompt_id: string | null;
  completed_prompt_ids: string[];
  issued: boolean;
};

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function isFalsy(value: unknown): boolean {
  // Python also treats empty lists and dictionaries as missing IDs.
  return (
    !value ||
    (Array.isArray(value) && value.length === 0) ||
    (isRecord(value) && Object.keys(value).length === 0)
  );
}

function isDigest(value: unknown): value is string {
  return (
    typeof value === "string" &&
    value.length === 64 &&
    /^[0-9a-f]{64}$/.test(value)
  );
}

function parseState(value: unknown): State {
  if (!isRecord(value) || Object.keys(value).length !== 3) {
    throw new Error("invalid state");
  }
  const current = value.current_prompt_id;
  const completed = value.completed_prompt_ids;
  const issued = value.issued;
  if (
    (current !== null && !isDigest(current)) ||
    !Array.isArray(completed) ||
    completed.length > 3 ||
    !completed.every(isDigest) ||
    typeof issued !== "boolean" ||
    issued !== (completed.length === 3) ||
    (issued && current !== null)
  ) {
    throw new Error("invalid state");
  }
  const ids = current === null ? completed : [...completed, current];
  if (new Set(ids).size !== ids.length) throw new Error("invalid state");
  return {
    current_prompt_id: current,
    completed_prompt_ids: completed,
    issued,
  };
}

async function main() {
  if (process.env.AUTO_IMPROVE_REVIEWER) return;
  let phase = "input";
  try {
    const decoder = new TextDecoder("utf-8", { fatal: true, ignoreBOM: true });
    const event: unknown = JSON.parse(decoder.decode(readFileSync(0)));
    if (!isRecord(event)) throw new Error("invalid event");
    const name = event.hook_event_name;
    if ((name !== "UserPromptSubmit" && name !== "Stop") || "agent_id" in event)
      return;
    const sessionId = event.session_id;
    const promptId = event.prompt_id;
    if (isFalsy(sessionId) || isFalsy(promptId)) return;
    if (typeof sessionId !== "string" || typeof promptId !== "string") {
      throw new Error("invalid IDs");
    }
    const active = "stop_hook_active" in event ? event.stop_hook_active : false;
    if (typeof active !== "boolean")
      throw new Error("invalid continuation flag");
    if (active) return;

    phase = "state";
    process.umask(0o077);
    const root = process.env.XDG_STATE_HOME || join(homedir(), ".local/state");
    if (!isAbsolute(root)) throw new Error("state root must be absolute");
    const directory = join(root, "auto-improve/claude");
    mkdirSync(directory, { recursive: true, mode: 0o700 });
    if (!sessionId.isWellFormed() || !promptId.isWellFormed()) {
      throw new Error("invalid ID encoding");
    }
    const sessionKey = createHash("sha256").update(sessionId).digest("hex");
    const promptKey = createHash("sha256").update(promptId).digest("hex");
    const statePath = join(directory, sessionKey + ".json");
    const { default: koffi } = await import("koffi");
    const native = koffi.load(
      process.platform === "darwin"
        ? "/usr/lib/libSystem.B.dylib"
        : "libc.so.6",
    );
    const flock = native.func("int flock(int fd, int operation)");
    const lock = openSync(join(directory, sessionKey + ".lock"), "a", 0o600);
    try {
      // LOCK_EX | LOCK_NB on macOS and glibc; closing the descriptor releases it.
      if (flock(lock, 2 | 4) !== 0) throw new Error("lock unavailable");
      let state: State;
      if (lstatSync(statePath, { throwIfNoEntry: false })) {
        state = parseState(JSON.parse(decoder.decode(readFileSync(statePath))));
      } else {
        if (name === "Stop") return;
        state = {
          current_prompt_id: null,
          completed_prompt_ids: [],
          issued: false,
        };
      }
      if (state.issued || state.completed_prompt_ids.includes(promptKey))
        return;
      if (name === "UserPromptSubmit") {
        if (state.current_prompt_id === promptKey) return;
        state.current_prompt_id = promptKey;
      } else {
        if (state.current_prompt_id !== promptKey) return;
        state.current_prompt_id = null;
        state.completed_prompt_ids.push(promptKey);
        state.issued = state.completed_prompt_ids.length === 3;
      }

      const temporary = join(
        directory,
        `${sessionKey}.${randomBytes(16).toString("hex")}.tmp`,
      );
      const target = openSync(temporary, "wx", 0o600);
      try {
        try {
          writeFileSync(target, JSON.stringify(state));
          fsyncSync(target);
        } finally {
          closeSync(target);
        }
        renameSync(temporary, statePath);
        const directoryFd = openSync(directory, "r");
        try {
          fsyncSync(directoryFd);
        } finally {
          closeSync(directoryFd);
        }
      } finally {
        rmSync(temporary, { force: true });
      }

      // Persist first: a crash before the spawn may skip a review, never repeat it.
      if (state.issued) {
        phase = "spawn";
        const cwdField = event.cwd;
        const cwd =
          typeof cwdField === "string" && !isFalsy(cwdField)
            ? cwdField
            : process.cwd();
        const transcriptField = event.transcript_path;
        const transcriptPath =
          typeof transcriptField === "string" && !isFalsy(transcriptField)
            ? transcriptField
            : "unknown";
        const reviewId = randomUUID();
        const child = spawn(
          "sh",
          [
            "-c",
            REVIEWER_SCRIPT,
            "sh",
            directory,
            reviewId,
            sessionKey,
            cwd,
            transcriptPath,
            new Date().toISOString(),
            buildPrompt(transcriptPath, cwd),
          ],
          {
            detached: true,
            stdio: "ignore",
            env: { ...process.env, AUTO_IMPROVE_REVIEWER: "1" },
            cwd,
          },
        );
        child.unref();
      }
    } finally {
      closeSync(lock);
    }
  } catch {
    try {
      writeSync(2, `auto-improve: ${phase} failed; no review.\n`);
    } catch {
      // A closed diagnostic stream must not turn a hook failure into a nonzero exit.
    }
  }
}

await main();
